src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingTests.cs

```diff
@@ -0,0 +1,1072 @@
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate* unmanaged<ValueType, void> cb)
+    {
+        {|CA1420:cb(default)|};
+    }
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task FunctionPointerWithUnmanagedValueTypeParameter_WithAutoLayoutField_Emits_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate* unmanaged<ValueType, void> cb)
+    {
+        {|CA1420:cb(default)|};
+    }
+}
+
+struct ValueType
+{
+    ValueType2 v;
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType2
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task ManagedFunctionPointerWithUnmanagedValueTypeParameter_WithAutoLayoutField_Does_Not_Emit_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate*<ValueType, void> cb)
+    {
+        cb(default);
+    }
+}
+
+struct ValueType
+{
+    ValueType2 v;
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType2
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task MarshalOffsetOf_Emits_Diagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public void Test()
+    {
+        IntPtr offset = {|CA1421:Marshal.OffsetOf(typeof(ValueType), ""field"")|};
+        IntPtr offsetGeneric = {|CA1421:Marshal.OffsetOf<ValueType>(""field"")|};
+    }
+}
+
+struct ValueType
+{
+    int field;
+}
+";
+            await VerifyCSAnalyzerAsync(source);
+        }
+
+        [Fact]
+        public async Task MarshalSizeOf_Emits_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public void Test<T, U>(System.Type type)
+        where U : unmanaged
+    {
+        object obj = default(ValueType);
+        int instanceSize = {|CA1421:Marshal.SizeOf(obj)|};
+        int size = {|CA1421:Marshal.SizeOf(typeof(ValueType))|};
+        int sizeGeneric = {|CA1421:Marshal.SizeOf<ValueType>()|};
+        int sizePassedInType = {|CA1421:Marshal.SizeOf(type)|};
+        int sizePassedInGeneric = {|CA1421:Marshal.SizeOf<T>()|};
+        int sizePassedInGenericUnmanaged = {|CA1421:Marshal.SizeOf<U>()|};
+    }
```

Comment:
This tests whether we'll crash on accessing the first argument or not.

>                     "SizeOf" => invocation.TargetMethod.IsGenericMethod || invocation.Arguments[0].Value is ITypeOfOperation { TypeOperand.IsUnmanagedType: true },

```suggestion
        int nonGenericNoParameter = Marshal.SizeOf();
    }
```

Consider also adding the test for `StructureToPtr`